### ISSUE TITLE
global hang up

### ISSUE THREAD

**Issue by nirs** [sustrik/libmill#62]
Runing this script simulating concurrent using turorial/step6 in expected way:

```
import socket
import time

def connect():
    s = socket.socket()
    s.connect(("127.0.0.1", 5555))
    return s

while 1:
    connections = [connect() for _ in range(10)]                                                                                                                                
    for s in connections:
        s.recv(100)
    for s in connections:
        s.sendall("Name\r\n")
    for s in connections:
        s.recv(100)
    for s in connections:
        s.close()
```

tutorial/step6 hang up after 20 connections, when the last connection is closed.

```
$ tutorial/step6
Total number of connections: 1
Active connections: 1
Failed connections: 0

Total number of connections: 2
Active connections: 2
Failed connections: 0

Total number of connections: 3
Active connections: 3
Failed connections: 0

Total number of connections: 4
Active connections: 4
Failed connections: 0

Total number of connections: 5
Active connections: 5
Failed connections: 0

Total number of connections: 6
Active connections: 6
Failed connections: 0

Total number of connections: 7
Active connections: 7
Failed connections: 0

Total number of connections: 8
Active connections: 8
Failed connections: 0

Total number of connections: 9
Active connections: 9
Failed connections: 0

Total number of connections: 10
Active connections: 10
Failed connections: 0

Total number of connections: 10
Active connections: 9
Failed connections: 0

Total number of connections: 10
Active connections: 8
Failed connections: 0

Total number of connections: 10
Active connections: 7
Failed connections: 0

Total number of connections: 10
Active connections: 6
Failed connections: 0

Total number of connections: 10
Active connections: 5
Failed connections: 0

Total number of connections: 10
Active connections: 4
Failed connections: 0

Total number of connections: 10
Active connections: 3
Failed connections: 0

Total number of connections: 10
Active connections: 2
Failed connections: 0

Total number of connections: 10
Active connections: 1
Failed connections: 0

Total number of connections: 10
Active connections: 0
Failed connections: 0

Total number of connections: 11
Active connections: 1
Failed connections: 0

Total number of connections: 12
Active connections: 2
Failed connections: 0

Total number of connections: 13
Active connections: 3
Failed connections: 0

Total number of connections: 14
Active connections: 4
Failed connections: 0

Total number of connections: 15
Active connections: 5
Failed connections: 0

Total number of connections: 16
Active connections: 6
Failed connections: 0

Total number of connections: 17
Active connections: 7
Failed connections: 0

Total number of connections: 18
Active connections: 8
Failed connections: 0

Total number of connections: 19
Active connections: 9
Failed connections: 0

panic: global hang-up
```

I replaced the mill_panic() with mill_assert() so we can inspect the process state when it panics:

```
Total number of connections: 114
Active connections: 4
Failed connections: 0

Total number of connections: 115
Active connections: 5
Failed connections: 0

Total number of connections: 116
Active connections: 6
Failed connections: 0

Total number of connections: 117
Active connections: 7
Failed connections: 0

Total number of connections: 118
Active connections: 8
Failed connections: 0

Total number of connections: 119
Active connections: 9
Failed connections: 0

Assert failed: 0 (poller.c:160)

Program received signal SIGABRT, Aborted.
0x000000305fa35877 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	  return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb) p goredump()

COROUTINE  state                                      current                                  created
------------------------------------------------------------------------------------------------------------------------
{0}        fdwait(-1)                                 tcp.c:178                                <main>
{1}        ready                                      tutorial/step6.c:41                      tutorial/step6.c:105
{112}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{113}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{114}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{115}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{116}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{117}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{118}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{119}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{120}      chs(<1>)                                   tutorial/step6.c:86                      tutorial/step6.c:111
{121}      chs(<1>)                                   ---                                      tutorial/step6.c:111

CHANNEL  msgs/max    senders/receivers                          refs  done  created
------------------------------------------------------------------------------------------------------------------------
<1>      0/0         s:{112},{113},{114},{115},{116},{117},{118},{119},{120},{121} 1     no    tutorial/step6.c:104

$1 = void
(gdb) p *mill_running
$2 = {state = MILL_CHS, u_ready = {item = {next = 0x0}}, u_fdwait = {item = {next = 0x0, prev = 0x0}, expiry = 128496030}, u_choose = {clauses = {first = 0x7ffff7e4bab0, 
      last = 0x7ffff7e4bab0}, othws = 0, available = 0}, ctx = {jbuf = {{__jmpbuf = {140737488346752, 6321267569117360643, 4198560, 140737488347248, 0, 0, 
          6321267569111069187, -6321285767971081725}, __mask_was_saved = 0, __saved_mask = {__val = {0 <repeats 16 times>}}}}}, valbuf = {ptr = 0x0, capacity = 128, 
    buf = '\000' <repeats 127 times>}, result = 1, cls = 0x0, debug = {item = {next = 0x0, prev = 0x7ffff7d8bee8}, id = 121, created = 0x40573c "tutorial/step6.c:111", 
    current = 0x4056ce "tutorial/step6.c:86"}}
```

**Comment by nirs** [comment]
Another run, now with a stacktrace:

```
Total number of connections: 15
Active connections: 5
Failed connections: 0

Total number of connections: 16
Active connections: 6
Failed connections: 0

Total number of connections: 17
Active connections: 7
Failed connections: 0

Total number of connections: 18
Active connections: 8
Failed connections: 0

Total number of connections: 19
Active connections: 9
Failed connections: 0

Assert failed: 0 (poller.c:160)

Program received signal SIGABRT, Aborted.
0x000000305fa35877 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	  return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
(gdb) bt
#0  0x000000305fa35877 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x000000305fa36f68 in __GI_abort () at abort.c:89
#2  0x00000000004038a5 in mill_wait () at poller.c:160
#3  0x0000000000402622 in mill_suspend () at cr.c:58
#4  0x0000000000402316 in mill_choose_wait () at chan.c:283
#5  0x00000000004023bc in mill_chs (ch=0x608030, val=0x7ffff7d4bb08, sz=4, current=0x4056ce "tutorial/step6.c:86") at chan.c:296
#6  0x00000000004013ce in dialogue (as=0x60bf80, ch=0x608030) at tutorial/step6.c:86
#7  0x000000000040171f in main (argc=1, argv=0x7fffffffe078) at tutorial/step6.c:111
(gdb) bt full
#0  0x000000305fa35877 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
        resultvar = 0
        pid = 1588
        selftid = 1588
#1  0x000000305fa36f68 in __GI_abort () at abort.c:89
        save_stage = 2
        act = {__sigaction_handler = {sa_handler = 0x0, sa_sigaction = 0x0}, sa_mask = {__val = {0, 140737351301632, 207758616374, 1, 0, 140737351254016, 207762779456, 
              140737488346752, 140737351301632, 4198560, 207758643541, 0, 207763695280, 0, 207766657536, 207766651328}}, sa_flags = -134428864, sa_restorer = 0x0}
        sigs = {__val = {32, 0 <repeats 15 times>}}
#2  0x00000000004038a5 in mill_wait () at poller.c:160
        fired = 32767
        rc = -8080
        i = 4198560
#3  0x0000000000402622 in mill_suspend () at cr.c:58
        __PRETTY_FUNCTION__ = "mill_suspend"
#4  0x0000000000402316 in mill_choose_wait () at chan.c:283
        uc = 0x7ffff7d4bd68
        it = 0x0
        cl = 0x7ffff7d4baa0
#5  0x00000000004023bc in mill_chs (ch=0x608030, val=0x7ffff7d4bb08, sz=4, current=0x4056ce "tutorial/step6.c:86") at chan.c:296
        cl = {epitem = {next = 0x0, prev = 0x7ffff7e0baa0}, chitem = {next = 0x0}, cr = 0x7ffff7d4bd40, ep = 0x608038, val = 0x7ffff7d4bb08, idx = 0, available = 0, used = 1}
#6  0x00000000004013ce in dialogue (as=0x60bf80, ch=0x608030) at tutorial/step6.c:86
        mill_val = 2
        deadline = 129357021
        inbuf = "Name", '\000' <repeats 180 times>, "\200\275\330\367\377\177\000\000\001\000\000\000\000\000\000\000"...
        sz = 13
        outbuf = "Hello, Name!\n", '\000' <repeats 242 times>
        rc = 13
#7  0x000000000040171f in main (argc=1, argv=0x7fffffffe078) at tutorial/step6.c:111
        mill_anchor = {-8304}
        mill_filler = <error reading variable mill_filler (Cannot access memory at address 0x7ffff7fd0000)>
        mill_sp = 0x7ffff7e4bd40
        as = 0x60ecc0
        port = 5555
        ls = 0x608010
        ch = 0x608030
``

**Comment by sustrik** [comment]
Fixed. Plase re-check and confirm.

**Comment by nirs** [comment]
Fixed, 1787370 connections without hang up :+1:
